ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,226 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/celestiaorg/nmt"
+
+	"go.opentelemetry.io/otel/trace"
+
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// TODO(@distractedm1nd) Find a better figure than NumWorkersLimit for this pool.
+// namespacePool is a worker pool responsible for the goroutines spawned by getLeavesByNamespace
+var namespacePool = workerpool.New(NumWorkersLimit)
+
+// GetSharesByNamespace walks the tree of a given root and returns its shares within the given namespace.ID.
+// If a share could not be retrieved, err is not nil, and the returned array contains nil shares in place of the shares it was unable to retrieve.
+func GetSharesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]Share, error) {
+	ctx, span := tracer.Start(ctx, "get-shares-by-namespace")
+	defer span.End()
+
+	leaves, err := getLeavesByNamespace(ctx, bGetter, root, nID, maxShares)
+	if err != nil && leaves == nil {
+		return nil, err
+	}
+
+	shares := make([]Share, len(leaves))
+	for i, leaf := range leaves {
+		if leaf != nil {
+			shares[i] = leafToShare(leaf)
+		}
+	}
+
+	return shares, err
+}
+
+// wrappedWaitGroup is needed because waitgroups do not expose their internal counter,
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	jobs    chan *job
```

Comment:
Maybe we can get away by sharing over chan _copy_ of the struct instead of _pointer_? It could be dangerous to have access by pointer to the same struct from different routines

Reply:
Each job is only accessed by a single goroutine, so it is safe here

Reply:
Extra point: As a value it can be stack allocated, thus reducing load on GC and heap size.
It will increase performance by small margin. So feel free to resolve

Reply:
@Wondertan  thoughts?

Reply:
Check out with `
go build -gcflags "-m" file.go
`

Reply:
It suggests if a value is allocated on the heap and other potential small optimisations

Reply:
Thanks! Didnt catch it, resolving